### PR TITLE
Update MLP.ipynb

### PR DESCRIPTION
**Pull request by arsaha28**

### CODE REVIEW

File: MLP.ipynb [hunk 1]
```diff
@@ -1,1146 +1,1188 @@
...
+    "    return accuracy, all_preds, all_labels\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "accuracy, all_preds, all_labels = evaluate_model(model, test_loader, device, nn_network, class_names)"
```
Comment (mohananup) [inline MLP.ipynb:1094]: we are returning the accuracy but not using ot, do you want to get use it for print the value outside or you will be using it in hyperparamter?

File: MLP.ipynb [hunk 2]
```diff
@@ -1,1146 +1,1188 @@
...
+    "- **Data Loading**: Loaded test data and moved it to the selected device.\n",
+    "- **Forward Passes**:\n",
+    "  - Passed data through the CNN and converted the output to a numpy array.\n",
+    "  - Passed the numpy array through the fully connected network to get predictions.\n",
+    "- **Accuracy Calculation**: Calculated accuracy by comparing predictions with true labels.\n",
+    "- **Results**:\n",
+    "  - Printed the test accuracy.\n",
+    "  - Generated and printed a classification report using `classification_report` from `sklearn`.\n",
+    "  - Created a confusion matrix and converted it to a DataFrame for better visualization.\n",
+    "- **Wandb Finishing**: Called `wandb.finish()` to complete the logging process."
```
Comment (mohananup) [inline MLP.ipynb]: this needs to be removed

File: MLP.ipynb [hunk 3]
```diff
@@ -1,1146 +1,1188 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "q8lI95KDnxBd"
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.metrics import confusion_matrix, classification_report\n",
```
Comment (mohananup) [inline MLP.ipynb]: we dont need to import all of them as it is already imported

File: MLP.ipynb [hunk 4]
```diff
@@ -1,1146 +1,1188 @@
...
+    "    return val_loss, val_accuracy"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_and_validate(model, train_loader, val_loader, nn_network, device, nc, learning_rate=0.0001, epochs=1):\n",
```
Comment (mohananup) [inline MLP.ipynb]: let's dont provide default paramter value for  learning_rate=0.0001, epochs=1

File: MLP.ipynb [hunk 5]
```diff
@@ -1,1146 +1,1188 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_accuracies, train_losses, val_accuracies, val_losses = train_and_validate(\n",
+    "    model, train_loader, val_loader, nn_network, device, nc, learning_rate=0.0001, epochs=1\n",
```
Comment (mohananup) [inline MLP.ipynb:999]: NIT: can we keep it in dict or as a variable so that functions are not disturbed.

File: MLP.ipynb
Comment (mohananup) [inline MLP.ipynb:1]: Need to remove reference of wandb from markdowns

**Review by mohananup — CHANGES_REQUESTED** [review]
some comments otherwise LGTM

File: MLP.ipynb [hunk 6]
```diff
@@ -1158,6 +1158,41 @@
...
+    "        # Update the best accuracy and parameters if the current model is better\n",
+    "        if max_val_accuracy > best_accuracy:\n",
+    "            best_accuracy = max_val_accuracy\n",
+    "            best_params = params\n",
+    "    return best_params, best_accuracy\n",
+    "\n",
+    "# Example usage:\n",
+    "param_grid = [\n",
+    "    {'epochs': 1, 'learning_rate': 0.01},\n",
+    "    {'epochs': 2, 'learning_rate': 0.001},\n",
```
Comment (mohananup) [inline MLP.ipynb:1183]: Here, In param grid, also we are hardcoding them only advantage is we can specify them together. 

Also, we need to reset the model and nn_network on every run, other wise the weights wont be re-initialised from the initial value.

**Review by mohananup — CHANGES_REQUESTED** [review]

File: MLP.ipynb (hunk 2 above)
Comment (arsaha28) [reply MLP.ipynb]: Done

File: MLP.ipynb
Comment (arsaha28) [reply MLP.ipynb:1]: Done

File: MLP.ipynb (hunk 6 above)
Comment (arsaha28) [reply MLP.ipynb:1183]: Done

**Review by mohananup — APPROVED** [review]
